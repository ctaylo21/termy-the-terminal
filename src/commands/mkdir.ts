import cloneDeep from 'lodash/cloneDeep';
import set from 'lodash/set';
import has from 'lodash/has';
import { getInternalPath } from './utilities/index';

/**
 * Given a folder path, creates that folder for the given file system and
 * returns the new file system if valid. If not, rejects with error.
 *
 * @param fileSystem {object} - filesystem to mkdir upon
 * @param currentPath {string} - current path within filesystem
 * @param folderPath  {string} - folder path to create
 * @returns Promise<object> - resolves with new path created if successful, rejects if not
 */
export default function mkdir(
  fileSystem: FileSystem,
  currentPath: string,
  folderPath: string,
): Promise<CommandResponse> {
  return new Promise((resolve, reject): void => {
    if (has(fileSystem, getInternalPath(currentPath, folderPath))) {
      reject('Path already exists');
    }

    const newFolder: TerminalFolder = {
      type: 'FOLDER',
      children: null,
    };

    const newFileSystem = cloneDeep(fileSystem);
    set(newFileSystem, getInternalPath(currentPath, folderPath), newFolder);

    resolve({
      commandResult: `Folder created: ${folderPath}`,
      updatedState: {
        fileSystem: newFileSystem,
      },
    });
  });
}
