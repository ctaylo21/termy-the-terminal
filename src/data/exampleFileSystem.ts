const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          video1: {
            type: 'FILE',
          },
        },
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
};

export default exampleFileSystem;
