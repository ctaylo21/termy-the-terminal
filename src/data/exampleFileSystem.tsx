import React from 'react';
import { FileSystem } from '../index';
import dogImg from '../../src/images/dog.png';

interface BlogPostProps {
  date: string;
  content: string;
}

export const BlogPost: React.FC<BlogPostProps> = ({
  date,
  content,
}): JSX.Element => (
  <>
    <h3>{date}</h3>
    <p>{content}</p>
  </>
);

const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          file2: {
            type: 'FILE',
            content: 'Contents of file 2',
            extension: 'txt',
          },
        },
      },
      dog: {
        type: 'FILE',
        content: dogImg,
        extension: 'png',
      },
      file1: {
        type: 'FILE',
        content: 'Contents of file 1',
        extension: 'txt',
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
  file3: {
    type: 'FILE',
    content: 'Contents of file 3',
    extension: 'txt',
  },
  blog: {
    type: 'FILE',
    content: <BlogPost date="3/22" content="Today is a good day" />,
    extension: 'txt',
  },
};

export default exampleFileSystem;
