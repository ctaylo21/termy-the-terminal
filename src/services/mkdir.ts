import cloneDeep from 'lodash/cloneDeep';
import set from 'lodash/set';
import { getInternalPath } from './utilities/index';
import has from 'lodash/has';

/**
 * Given a folder path, creates that folder for the given file system and
 * returns the new file system if valid. If not, rejects with error.
 *
 * @param fileSystem {object} - filesystem to mkdir upon
 * @param currentPath {string} - current path within filesystem
 * @param folderPath  {string} - folder path to create
 * @returns Promise<object> - resolves with new path created if successful, rejects if not
 */
export default function mkdir(
  fileSystem: FileSystem,
  currentPath: string,
  folderPath: string,
): Promise<ServiceResponse> {
  return new Promise(
    (resolve, reject): void => {
      if (has(fileSystem, getInternalPath(currentPath, folderPath))) {
        reject('Path already exists');
      }

      const newFolder: TerminalFolder = {
        type: 'FOLDER',
        children: null,
      };

      const newFileSystem = cloneDeep(fileSystem);
      set(newFileSystem, getInternalPath(currentPath, folderPath), newFolder);

      resolve({
        serviceResult: `Folder created: ${folderPath}`,
        updatedState: {
          fileSystem: newFileSystem,
        },
      });
    },
  );
}
