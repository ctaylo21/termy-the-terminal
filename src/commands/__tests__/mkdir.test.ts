import mkdir from '../mkdir';
import cloneDeep from 'lodash/cloneDeep';
import set from 'lodash/set';
import exampleFileSystem from '../../data/exampleFileSystem';

describe('mkdir suite', (): void => {
  describe('success', (): void => {
    it('should create directory from root', async (): Promise<
      CommandResponse
    > => {
      const expectedFileSystem = cloneDeep(exampleFileSystem);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      set(expectedFileSystem, 'newDir', {
        type: 'FOLDER',
        children: null,
      });

      return expect(
        mkdir(exampleFileSystem, '/', 'newDir'),
      ).resolves.toStrictEqual({
        commandResult: 'Folder created: newDir',
        updatedState: {
          fileSystem: expectedFileSystem,
        },
      });
    });

    it('should create directory from nested path', async (): Promise<
      CommandResponse
    > => {
      const expectedFileSystem = cloneDeep(exampleFileSystem);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      set(expectedFileSystem, 'home.children.newDir', {
        type: 'FOLDER',
        children: null,
      });

      return expect(
        mkdir(exampleFileSystem, '/home', 'newDir'),
      ).resolves.toStrictEqual({
        commandResult: 'Folder created: newDir',
        updatedState: {
          fileSystem: expectedFileSystem,
        },
      });
    });

    it('should create directory given a folder path', async (): Promise<
      CommandResponse
    > => {
      const expectedFileSystem = cloneDeep(exampleFileSystem);

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      set(expectedFileSystem, 'home.children.newDir', {
        type: 'FOLDER',
        children: null,
      });

      return expect(
        mkdir(exampleFileSystem, '/', 'home/newDir'),
      ).resolves.toStrictEqual({
        commandResult: 'Folder created: home/newDir',
        updatedState: {
          fileSystem: expectedFileSystem,
        },
      });
    });
  });

  describe('Failure', (): void => {
    it('should reject if path already exists', async (): Promise<
      CommandResponse
    > => {
      return expect(
        mkdir(exampleFileSystem, '/home', 'user'),
      ).rejects.toMatchSnapshot();
    });
  });
});
