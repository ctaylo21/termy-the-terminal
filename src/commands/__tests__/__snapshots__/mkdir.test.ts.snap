// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mkdir suite Failure should reject if path already exists 1`] = `"Path already exists"`;
